README.md

```diff
@@ -26,6 +26,7 @@ Run `make` to build `./bin/cloudwatch.so`. Then use with Fluent Bit:
 * `log_group_name`: The name of the CloudWatch Log Group that you want log records sent to.
 * `log_stream_name`: The name of the CloudWatch Log Stream that you want log records sent to.
 * `log_stream_prefix`: Prefix for the Log Stream name. The tag is appended to the prefix to construct the full log stream name. Not compatible with the `log_stream_name` option.  
+* `log_stream_key_name`: If provided, this key name is extracted from the log record and placed into the log stream name. The value must be a string, and will be suffixed to `log_stream_prefix` and the `tag`. If the key is not found in the log then `log_stream_prefix` or `log_stream_name` (whichever is provided) will be used instead.
```

Comment:
@davidnewhall What is the logic behind making the log stream name: 

```
log_stream_prefix + tag + log_stream_key_name
```

Without some justification, at first glance I'd prefer if the field works the same was as it does in the Fluentd CloudWatch Logs plugin- the value of the key specified with `log_stream_key_name` becomes the log stream name.

Later this year we plan to work on migrating the recommended CW Logs EKS daemonset to use Fluent Bit instead of Fluentd, I suspect that will be easier if the log_stream_key_name feature works the same way: https://github.com/aws-samples/amazon-cloudwatch-container-insights/blob/master/k8s-deployment-manifest-templates/deployment-mode/daemonset/container-insights-monitoring/fluentd/fluentd.yaml#L176

Reply:
This explanation isn't entirely accurate, and needs to be fixed. If `log_stream_prefix` is empty then the stream name becomes `tag + log_stream_key_name`. I personally like having the `tag` added to the log stream name, but I realize some folks don't care for that. Would another bool variable be OK to trigger that? Something like `prefix_log_stream_with_tag`.

If you don't want _the_ prefix, don't set the `log_stream_prefix` variable. The `log_stream_name` value will only be used if the provided key is missing from the log line, or empty. Does this seem OK?

Take a look at the `switch/case` statement: https://github.com/aws/amazon-cloudwatch-logs-for-fluent-bit/pull/78/files#diff-c88e64246a5b33fe05db284d0bec2b45R371-R380

I'll update the code and the explanation in the readme with your next round of feedback for the questions above.

Thank you!

Reply:
I edited the README locally, and changed the line to this: 
* `log_stream_key_name`: If provided, this key name is extracted from the log record and placed into the log stream name. The value must be a string, and will be suffixed to `log_stream_prefix` and the `tag`. Leave `log_stream_prefix` empty to avoid having that prefix added. If the key is not found in the log then `log_stream_prefix` or `log_stream_name` (whichever is provided) will be used instead.

I added this sentence: 
> Leave log_stream_prefix empty to avoid having that prefix added.

My open questions:
- Is adding `tag + "."` as a prefix to `log_stream_key_name ` OK? I think this is what you actually have a problem with, so I'm assuming this is not OK. Next questions:
- Is adding a boolean input parameter to turn on the `tag` prefix OK? I personally want the tag in the name.
- If yes, should that parameter also control the situation where `log_stream_key_name` is not set and `log_stream_prefix` is set?

Reply:
> I think this is what you actually have a problem with, so I'm assuming this is not OK. Next questions:

Yeah it is. I want the default behavior to be the same as the Fluentd plugin- the value of the key becomes the log stream name.

> Leave log_stream_prefix empty to avoid having that prefix added.

I think this is fine, since you have documented it.

> Something like prefix_log_stream_with_tag

That would be an acceptable way to do it IMO. I think the name might be confusing though, since it's not clear anymore which is the "first" prefix. The tag (which this parameter says is a prefix) or the log stream prefix. Not sure of a better name though.

Just to make things clear; below I outline what I'm alright with (and another option which would be ideal but would be a ton of work).

### Proposed Solution

- the value of the key specified with `log_stream_key_name` becomes the stream name. If you have `log_stream_prefix` set, that prefix is added. The tag is not added at all, and the prefix is optional.
- Something like `prefix_log_stream_with_tag` adds the tag in there. So you can have `log_stream_prefix.tag.log_stream_key`. I think dots is okay as delimiter but you could also use dashes. Not sure which is more typical for cloudwatch (I don't think there is a pattern/recommendation).

### Ideal solution

The new rewrite tag plugin in fluent bit core introduced a syntax for selecting keys and parts of the tag: https://docs.fluentbit.io/manual/pipeline/filters/rewrite-tag

We would add a new field like `log_stream_format` that uses that syntax. The value of that option might be:

```
log_stream_format   logs-$TAG[1]-$stage
```

If the tag is `logs.my-app-name` and the log record looks like what is shown below:

```
{
     "log": "some message",
     "stage": "prod",
}
```

Then the log stream name would be: `logs-my-app-name-prod`

That would be really cool and is the ideal solution IMO. I am not expecting you to build this though. I may eventually add it to the new core fluent bit cloudwatch plugin, since it can re-use the code from rewrite tag: https://docs.fluentbit.io/manual/pipeline/outputs/cloudwatch

Reply:
As I was reading the first half of your comment, I was leaning toward "we could use a template for the tag name instead of arbitrary string concatenation" and then you right in and said the same thing. I'll see what I can cook up. Thanks for your feedback!

Reply:
Added templating. Might not be exactly what you were thinking, but it works pretty well. The test procedures probably provide the best documentation of the behavior. More feedback welcomed! I still haven't tested this live, but will shortly.

Reply:
I extended this same approach to log group names in another branch. I'll do more testing on that, but it should work very similarly.